pallets/ethereum-transaction/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+		///
+		/// NOTE - The execution fees are charged by the Ethereum pallet,
+		/// we only have to charge for the nonce read operation.
+		fn call(
+			from: H160,
+			to: H160,
+			data: &[u8],
+			value: U256,
+			gas_price: U256,
+			gas_limit: U256,
+		) -> DispatchResultWithPostInfo {
+			let nonce = Nonce::<T>::get();
+			let read_weight = T::DbWeight::get().reads(1);
+
+			let signature =
+				Pallet::<T>::get_transaction_signature().ok_or(DispatchErrorWithPostInfo {
+					post_info: PostDispatchInfo {
+						actual_weight: Some(read_weight),
+						pays_fee: Pays::Yes,
+					},
+					error: DispatchError::Other("Failed to create transaction signature"),
+				})?;
+
+			let transaction = TransactionV2::Legacy(LegacyTransaction {
+				nonce,
+				gas_price,
+				gas_limit,
+				action: TransactionAction::Call(to),
+				value,
+				input: data.into(),
+				signature,
+			});
+
+			Nonce::<T>::put(nonce.saturating_add(U256::one()));
+
+			let (_target, _value, info) = pallet_ethereum::Pallet::<T>::execute(
+				from,
+				&transaction,
+				Some(T::config().clone()),
+			)
+			.map_err(|e| {
+				let weight = e.post_info.actual_weight.map_or(Weight::zero(), |w| w);
+
+				DispatchErrorWithPostInfo {
+					post_info: PostDispatchInfo {
+						actual_weight: Some(weight.saturating_add(read_weight)),
+						pays_fee: Pays::Yes,
+					},
+					error: e.error,
+				}
+			})?;
+			match info {
+				CallOrCreateInfo::Call(call_info) => {
+					Self::deposit_event(Event::Executed {
+						from,
+						to,
+						exit_reason: call_info.exit_reason.clone(),
+						value: call_info.value.clone(),
+					});
+
+					match call_info.exit_reason {
+						ExitReason::Succeed(_) => Ok(PostDispatchInfo {
+							actual_weight: Some(read_weight),
+							pays_fee: Pays::Yes,
+						}),
+						ExitReason::Error(e) => Err(DispatchErrorWithPostInfo {
+							post_info: PostDispatchInfo {
+								actual_weight: Some(read_weight),
+								pays_fee: Pays::Yes,
+							},
+							error: map_evm_error(e),
+						}),
+						ExitReason::Revert(_) => Err(DispatchErrorWithPostInfo {
+							post_info: PostDispatchInfo {
+								actual_weight: Some(read_weight),
+								pays_fee: Pays::Yes,
+							},
+							error: DispatchError::Other("EVM encountered an explicit revert"),
+						}),
+						ExitReason::Fatal(e) => Err(DispatchErrorWithPostInfo {
+							post_info: PostDispatchInfo {
+								actual_weight: Some(read_weight),
+								pays_fee: Pays::Yes,
+							},
+							error: map_evm_fatal_error(e),
+						}),
+					}
+				}
+				CallOrCreateInfo::Create(_) => Err(DispatchErrorWithPostInfo {
+					post_info: PostDispatchInfo {
+						actual_weight: Some(read_weight),
+						pays_fee: Pays::Yes,
+					},
+					error: DispatchError::Other("unexpected execute result"),
+				}),
+			}
+		}
+	}
+
+	fn map_evm_error(e: ExitError) -> DispatchError {
+		match e {
+			ExitError::StackUnderflow => DispatchError::Other("stack underflow"),
```

Comment:
Hmm, initially I added this `DispatchError` because I wasn't able to debug test failure but completely forgot about the "outside". Awesome point, will definitely change this, thank you!

Reply:
Done. I'm not passing the error message in the case of `ExitError::Other(Cov<'static, str>)` though. It seems that might be a bit tricky to do if we really want to, WDYT?